File: game-server/src/main/java/work/fking/pangya/game/model/IffContainer.java
package work.fking.pangya.game.model;

import java.util.List;

public interface IffContainer<I extends IffObject> {

    List<I> entries();

    default I findByIffId(int iffId) {
        for (var entry : entries()) {
            if (entry.iffId() == iffId) {
                return entry;
            }
        }
        return null;
    }

    default I findByUid(int uid) {
        for (var entry : entries()) {
            if (entry.uid() == uid) {
                return entry;
            }
        }
        return null;
    }

    default int size() {
        return entries().size();
    }

    default List<I> chunk(int offset, int size) {
        var entries = entries();
        return entries.subList(offset, Math.min(size, entries.size()));
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/model/IffContainer.kt
package work.fking.pangya.game.model

interface IffContainer<I : IffObject> {
    val entries: List<I>

    fun findByIffId(iffId: Int): I? {
        return entries.find { it.iffId == iffId }
    }

    fun findByUid(uid: Int): I? {
        return entries.find { it.uid == uid }
    }

    fun existsByUid(uid: Int): Boolean {
        return findByUid(uid) != null
    }
}
